### PR TITLE
Skaffold enhancements

### PR DESCRIPTION
**Pull request by briantopping**
**How to categorize this PR?**
/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Out of box, the Gardener build cannot use `make gardener-debug` or be plugged into an IDE that supports the [Google Cloud Code](https://cloud.google.com/code) plugin. The benefit of this plugin is the automation of connections between the debugger and the Delve debugger nubs that are attached as sidecars to the deployed Gardener artifacts. This PR fixes both.

This patch enables this in a few ways:
* The deployment for admission-controller is updated so the health check ports are exposed. This is necessary for Skaffold to remap the port exposures when it deploys the app with Delve.
* Allow enablement of health checks via `ENABLE_HEALTH_CHECKS` environment variable. When boot complexity causes some components to be loaded before all of its dependencies, the components tend to accept their state and report unhealthy rather than restart themselves. Without the health probe watchdog from k8s, the controllers will remain stuck in that state, even as the necessary resources are eventually added. The singular health check disable is easier to manage and document.
* Remove the `BUILD_DATE` injection from the Makefile to Skaffold via environment variables. This is replaced by Go templates in the Skaffold config, enabled by https://github.com/GoogleContainerTools/skaffold/pull/9005 and released in Skaffold v2.7.0. In doing so, there are only static environment variables that must be encoded in the Google Cloud Code plugin. These do not need to be changed with build iterations.
* Provide documentation. The documentation includes information about how to set up Google Cloud Code and the Makefile now emits the necessary static environment variables to do so. In case the variables change, this reduces the burden on future documentation updates. 

**Release note**:

```noteworthy developer
Enhance Skaffold experience to make `make gardener-debug` work out-of-box and also be usable from within Google Cloud Code IDE plugins.
```

### CODE REVIEW

**Comment by briantopping** [comment]
Hi @oliver-goetz, I wonder if I might get your advice on this PR. Hopefully the description makes sense and the effort resonates with your very valuable basis here. I suspect one problem may be that the helm values files changed and they are [not working](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/8418/pull-gardener-unit/1697726802073489408) with the upgrade tests from 1.78. But I'm also a bit overwhelmed with information here and not sure why some things seem to fail but are not considered failures.

File: Makefile [hunk 1]
```diff
@@ -45,6 +44,7 @@ IPFAMILY                                   := ipv4
 PARALLEL_E2E_TESTS                         := 15
 GARDENER_RELEASE_DOWNLOAD_PATH             := $(REPO_ROOT)/dev
 PRINT_HELP ?=
+SKAFFOLD_DISABLE_HEALTH_CHECKS ?=
```
Comment (oliver-goetz) [inline Makefile]: From my perspective this option complicates things unnecessarily. Debugging without disabling the health checks does not really make sense, because the components always die.
The rationale of debugging is to bring up the local setup first and then start the components you plan to debug in debugging mode.
There is an example for this in the [docs](https://github.com/gardener/gardener/blob/master/docs/deployment/getting_started_locally.md#debugging-gardener)
```
# initial deployment of all components
make gardener-up
# start debugging gardenlet without deploying other components
make gardener-debug SKAFFOLD_MODULE=gardenlet
``` 
This works with a single gardener component as well as for multiple components.

With this new env it would be more complicated.
```
make gardener-up
make gardener-debug SKAFFOLD_MODULE=gardenlet SKAFFOLD_DISABLE_HEALTH_CHECKS=gardenlet
```
Starting the local setup with `make gardener-debug` is not intended, so I don't see a reason to make it work, when I see the negative side effects of the change. Even when you plan to debug every gardener component at the same time you could run `make kind-up gardener-up gardener-debug` which brings everything up without restarting any pods manually.

File: charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml [hunk 2]
```diff
@@ -51,8 +51,8 @@ spec:
           type: RuntimeDefault
       containers:
       - name: gardener-resource-manager
-        image: {{ .Values.global.image.repository }}:{{ .Values.global.image.tag }}
-        imagePullPolicy: {{ .Values.global.image.pullPolicy }}
+        image: {{ .Values.global.resourceManager.image.repository }}:{{ .Values.global.image.tag }}
+        imagePullPolicy: {{ .Values.global.resourceManager.image.pullPolicy }}
         args:
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml]: why is this needed? Afaik the resource-manager helm chart is not even used by the local setup, but resource-manager is deployed by `gardenlet` using the [resource-manager go component](https://github.com/gardener/gardener/tree/master/pkg/component/resourcemanager).

File: charts/gardener/resource-manager/values.yaml [hunk 3]
```diff
@@ -1,8 +1,9 @@
 global:
-  image:
-    repository: eu.gcr.io/gardener-project/gardener/gardener-resource-manager
-    tag: latest
-    pullPolicy: IfNotPresent
+  resourceManager:
+    image:
+      repository: eu.gcr.io/gardener-project/gardener/gardener-resource-manager
+      tag: latest
+      pullPolicy: IfNotPresent
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/values.yaml]: same here

File: skaffold.yaml [hunk 4]
```diff
@@ -514,19 +514,19 @@ profiles:
...
   - op: add
-    path: /deploy/helm/releases/0/setValues
+    path: /deploy/helm/releases/0/setValueTemplates
     value:
       global.apiserver.replicaCount: 1
       global.apiserver.livenessProbe.enable: false
       global.apiserver.readinessProbe.enable: false
       global.admission.replicaCount: 1
-      global.admission.config.server.healthProbes.enable: false
+      global.admission.config.server.healthProbes.enable: "{{ default \"\" .SKAFFOLD_DISABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"admission\" | ternary \"false\" \"true\" }}"
```
Comment (oliver-goetz) [inline skaffold.yaml]: Please see my first comment about `SKAFFOLD_DISABLE_HEALTH_CHECKS` env.

File: skaffold.yaml [hunk 5]
```diff
@@ -1092,6 +1092,8 @@ deploy:
       setValueTemplates:
         image.repository: '{{.IMAGE_REPO_eu_gcr_io_gardener_project_gardener_gardenlet}}'
         image.tag: '{{.IMAGE_TAG_eu_gcr_io_gardener_project_gardener_gardenlet}}@{{.IMAGE_DIGEST_eu_gcr_io_gardener_project_gardener_gardenlet}}'
+        global.resourceManager.image.repository: '{{.IMAGE_REPO_eu_gcr_io_gardener_project_gardener_resource_manager}}'
+        global.resourceManager.image.tag: '{{.IMAGE_TAG_eu_gcr_io_gardener_project_gardener_resource_manager}}@{{.IMAGE_DIGEST_eu_gcr_io_gardener_project_gardener_resource_manager}}'
```
Comment (oliver-goetz) [inline skaffold.yaml]: I'm not sure what this is supposed to do. There is no `global.resourceManager.image.repository` tag in gardenlet helm chart and the image of `gardener-resource-manager` is already overwritten using `imageVectorOverwrite` 4 lines below.

File: skaffold.yaml [hunk 6]
```diff
@@ -1183,8 +1185,8 @@ profiles:
   - command: debug
   patches:
   - op: add
-    path: /deploy/helm/releases/0/setValues
+    path: /deploy/helm/releases/0/setValueTemplates
     value:
       replicaCount: 1
       config.leaderElection.leaderElect: false
-      config.server.healthProbes.enable: false
+      config.server.healthProbes.enable: "{{ default \"\" .SKAFFOLD_DISABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"gardenlet\" | ternary \"false\" \"true\" }}"
```
Comment (oliver-goetz) [inline skaffold.yaml]: Please see my first comment about `SKAFFOLD_DISABLE_HEALTH_CHECKS` env.

**Review by oliver-goetz — COMMENTED** [review]
Thanks for improving our debugging capabilities 😄 

The upgrade tests fail during helm install of gardener-controlplane.
Thus, I suggest to debug it locally, by checking out gardener the previous release branch (`release-v1.78`) first and running `make kind-up gardener-up`. Then, please check out `master` branch and finally run `make gardener-up` again.
When running the final `make gardener-up` you might be able to find the reason for the failing tests.

I have some remarks about your changes too.

**Comment by briantopping** [comment]
Thanks for your review, Oliver!  I appreciate you finding the Helm problem, as I mentioned in Slack I was unaware of the Prow artifact logs. I will get that resolved so it is easier to have a conversation about the other issues. 

The intent of this change is to enable the Google Cloud Code plugin, these are the changes that are required to make it happen. It seems most of your concerns are about the new means to enable or disable health checks. I look forward to convincing you that this enhances the developer experience rather than makes it more complex.

We have our version of the May 1 holiday here, so please look forward to my response at the start of your day on Wednesday!

File: Makefile (hunk 1 above)
Comment (briantopping) [reply Makefile]: Yes, I see what you are saying here. I wonder if it would feel less complex if the `SKAFFOLD_MODULE` and `SKAFFOLD_DISABLE_HEALTH_CHECKS` variables were merged? In other words, the module being debugged was automatically disabled. 

Regarding the feature and configuration in general, of course there are many paths to learning a code base like Gardener. Documentation usually trails the code for accuracy, so it's good practice to reduce onboarding documentation for new users to a few steps that will practically never change. I agree that "starting the local setup with make gardener-debug is not intended" in the current configuration, so I am not surprised that it's challenging to do it. 

Of course this could be tested and validated more easily than this PR, but the goal of this PR is to load Gardener from an IDE running Google Cloud Code ("GCC"), not run it from the CLI. Skaffold at the CLI emits debugger ports, but it does not *do* anything with them. Were the user to connect to these ports, they may change with successive iterations. Using GCC to launch Skaffold this automates all this administrative detail and significantly simplifies the user experience of debugging Gardener. 

Most importantly, I am advocating as a part of this that whatever increase in complexity that this PR introduces, the reduction in complexity for debugging far outweighs it.

File: skaffold.yaml (hunk 4 above)
Comment (briantopping) [reply skaffold.yaml]: May we merge this into the discussion above?

File: skaffold.yaml (hunk 6 above)
Comment (briantopping) [reply skaffold.yaml]: May we merge this into the discussion above?

While this template looks complex, all it's doing is creating a truthy value on whether the string contains the particular component name. This would be a trivial reference if there was a different environment variable for each component, but the user experience complexity of this is reduced if the developer can look at a single environment variable to see what is enabled instead of several. I am working with the Skaffold team on global templates within a configuration, all the new template instantiations would become very short with that in place.

**Comment by briantopping** [comment]
Thanks again for the response, @oliver-goetz! You filled in a huge gap for me about how the build and CI works! 👍 

To reiterate what I said above, the goal here isn't to make the CLI experience better, it's to make iterative development in the IDE with Google Cloud Code (the companion IDE plugin for Skaffold) simple and productive.

File: charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml (hunk 2 above)
Comment (briantopping) [reply charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml]: Artifact of something else I broke, thanks for catching that!

File: charts/gardener/resource-manager/values.yaml (hunk 3 above)
Comment (briantopping) [reply charts/gardener/resource-manager/values.yaml]: Artifact of something else I broke, thanks for catching that!

File: skaffold.yaml (hunk 5 above)
Comment (briantopping) [reply skaffold.yaml]: Artifact of something else I broke, thanks for catching that!

File: docs/deployment/getting_started_locally.md [hunk 7]
```diff
@@ -136,11 +136,51 @@ Please check your console output for the concrete port-forwarding on your machin
...
+
+Thus, new developers are advised to set `ENABLE_HEALTH_CHECKS`, both at the command line with the various make environments and with Google Cloud Code IDE plugin. This will cause health checks to be deployed by Helm as they would be in
+production. Note this is **_not_** a good long term debugging configuration though.
+
+```bash
+make gardener-debug ENABLE_HEALTH_CHECKS=admission,controller,scheduler,gardenlet,operator
+```
+
+Once a developer discovers a health check disturbing their debugging workflow, they should remove the check from the `ENABLE_HEALTH_CHECKS` list. The available checks are currently from the set
+of `{admission,controller,scheduler, gardenlet, operator}`.
```
Comment (oliver-goetz) [inline docs/deployment/getting_started_locally.md:160]: In my opinion the `ENABLE_HEALTH_CHECKS` environment variable is not useful. Debugging with enabled health checks does not work, because all go-routines are stopped at break points and the pods will be killed quite fast.
In practice nobody would use the parameter and in my opinion it is not a good idea to overload the skaffold setup with "once in a lifetime switches" for demo purposes only.
Additionally, I don't want to advice new developers to test non-working configurations, because this just annoys people.

File: skaffold-operator.yaml [hunk 8]
```diff
@@ -294,8 +294,8 @@ profiles:
   - command: debug
   patches:
   - op: add
-    path: /deploy/helm/releases/0/setValues
+    path: /deploy/helm/releases/0/setValueTemplates
     value:
       replicaCount: 1
       config.leaderElection.leaderElect: false
-      config.server.healthProbes.enable: false
+      config.server.healthProbes.enable: "{{ default \"\" .ENABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"operator\" }}"
```
Comment (oliver-goetz) [inline skaffold-operator.yaml]: Please see my first `ENABLE_HEALTH_CHECKS` comment.

File: skaffold.yaml [hunk 9]
```diff
@@ -514,19 +514,19 @@ profiles:
...
-      global.admission.config.server.healthProbes.enable: false
+      global.admission.config.server.healthProbes.enable: "{{ default \"\" .ENABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"admission\" }}"
       global.controller.replicaCount: 1
       global.controller.config.leaderElection.leaderElect: false
-      global.controller.config.server.healthProbes.enable: false
+      global.controller.config.server.healthProbes.enable: "{{ default \"\" .ENABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"controller\" }}"
       global.scheduler.replicaCount: 1
       global.scheduler.config.leaderElection.leaderElect: false
-      global.scheduler.config.server.healthProbes.enable: false
+      global.scheduler.config.server.healthProbes.enable: "{{ default \"\" .ENABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"scheduler\" }}"
```
Comment (oliver-goetz) [inline skaffold.yaml]: Please see my first `ENABLE_HEALTH_CHECKS` comment.

File: skaffold.yaml [hunk 10]
```diff
@@ -1183,8 +1183,8 @@ profiles:
   - command: debug
   patches:
   - op: add
-    path: /deploy/helm/releases/0/setValues
+    path: /deploy/helm/releases/0/setValueTemplates
     value:
       replicaCount: 1
       config.leaderElection.leaderElect: false
-      config.server.healthProbes.enable: false
+      config.server.healthProbes.enable: "{{ default \"\" .ENABLE_HEALTH_CHECKS | lower | splitList \",\" | has \"gardenlet\" }}"
```
Comment (oliver-goetz) [inline skaffold.yaml]: Please see my first `ENABLE_HEALTH_CHECKS` comment.

**Review by oliver-goetz — CHANGES_REQUESTED** [review]
This is the summary of our slack conversation about the `ENABLE_HEALTH_CHECKS` environment variable.

**Comment by briantopping** [comment]
Thanks Oliver. As you know, my response to this is that we have found the Gardener setup experience to still be far too difficult for developers who must learn the product in isolation. If Gardener is to thrive, we must reduce the cognitive load so that the majority of those who try are initially successful. 

I'm quite surprised at the resistance to making the system easier for new developers to use. Documentation is frequently insufficient in open source projects and nothing will change that. Expecting developers to know which pods to manually restart before they have even seen it running or having the lived experience of why to restart those pods is setting them up to fail.

This PR sidesteps those problems. With very little additional code and a single environment variable that is otherwise hidden from existing developers. These new developers can quickly load the entire environment, set quick breakpoints to build confidence in a running system, **_THEN_** learn how and why to disable health checks.

I sense that you are very disappointed by this request and my summary of our conversation is that I confirmed you would not be disappointed by us building a bigger consensus around the change. Let's let others from the team weigh in on this before spoiling the benefits of this effort with additional negativity.

File: skaffold-operator.yaml (hunk 8 above)
Comment (briantopping) [reply skaffold-operator.yaml]: Please see general comments.

File: skaffold.yaml (hunk 9 above)
Comment (briantopping) [reply skaffold.yaml]: Please see general comments.

File: skaffold.yaml (hunk 10 above)
Comment (briantopping) [reply skaffold.yaml]: Please see general comments.

File: docs/deployment/getting_started_locally.md (hunk 7 above)
Comment (briantopping) [reply docs/deployment/getting_started_locally.md:160]: Please see general comments.

**Comment by briantopping** [comment]
I will continue to spend time updating this until it is resolved. 

Regarding the introduction of a new flag and the contention that this introduces new flags, a reviewer should note it also removes a flag from the dev CLI UX, so there is net zero new flags. The new flag defaults are exactly the same as the current setup. In effect, there is zero net change unless a user needs to use this functionality, which I and others are waiting on.

**Review by briantopping — COMMENTED** [review]
I don't see any problems here.
